clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.cpp

```diff
@@ -0,0 +1,214 @@
+      RawCountingEnumPrefixes(Options.get(CountingEnumPrefixesOptionName, "")),
+      RawCountingEnumSuffixes(
+          Options.get(CountingEnumSuffixesOptionName, "count")),
+      ParsedCountingEnumPrefixes(
+          utils::options::parseStringList(RawCountingEnumPrefixes)),
+      ParsedCountingEnumSuffixes(
+          utils::options::parseStringList(RawCountingEnumSuffixes)),
+      CountingEnumPrefixesSet(
+          Options.get(CountingEnumPrefixesOptionName).has_value()),
+      CountingEnumSuffixesSet(
+          Options.get(CountingEnumSuffixesOptionName).has_value()),
+      CountingEnumConstantDecl(nullptr) {
+  if (!EnableCountingEnumHeuristic) {
+    if (CountingEnumPrefixesSet)
+      configurationDiag("%0: Counting enum heuristic is disabled but "
+                        "CountingEnumPrefixes is set")
+          << Name;
+    if (CountingEnumSuffixesSet)
+      configurationDiag("%0: Counting enum heuristic is disabled but "
+                        "CountingEnumSuffixes is set")
+          << Name;
+  }
+}
+
+void TaggedUnionMemberCountCheck::storeOptions(
+    ClangTidyOptions::OptionMap &Opts) {
+  Options.store(Opts, StrictModeOptionName, StrictMode);
+  Options.store(Opts, EnableCountingEnumHeuristicOptionName,
+                EnableCountingEnumHeuristic);
+  Options.store(Opts, CountingEnumPrefixesOptionName, RawCountingEnumPrefixes);
+  Options.store(Opts, CountingEnumSuffixesOptionName, RawCountingEnumSuffixes);
+}
+
+void TaggedUnionMemberCountCheck::registerMatchers(MatchFinder *Finder) {
+  Finder->addMatcher(
+      recordDecl(
+          anyOf(isStruct(), isClass()),
+          has(fieldDecl(hasType(qualType(hasCanonicalType(recordType()))))
+                  .bind("union")),
+          has(fieldDecl(hasType(qualType(hasCanonicalType(enumType()))))
+                  .bind("tags")))
+          .bind("root"),
+      this);
+}
+
+static bool isUnion(const FieldDecl *R) {
+  return R->getType().getCanonicalType().getTypePtr()->isUnionType();
+}
+
+static bool isEnum(const FieldDecl *R) {
+  return R->getType().getCanonicalType().getTypePtr()->isEnumeralType();
+}
+
+static bool hasMultipleUnionsOrEnums(const RecordDecl *Rec) {
+  return llvm::count_if(Rec->fields(), isUnion) > 1 ||
+         llvm::count_if(Rec->fields(), isEnum) > 1;
+}
+
+static bool signEquals(const llvm::APSInt &A, const llvm::APSInt &B) {
+  return (A.isNegative() && B.isNegative()) ||
+         (A.isStrictlyPositive() && B.isStrictlyPositive()) ||
+         (A.isZero() && B.isZero());
+}
+
+static bool greaterBySign(const llvm::APSInt &A, const llvm::APSInt &B) {
+  return (A.isNonNegative() && B.isNegative()) ||
+         (A.isStrictlyPositive() && B.isNonPositive());
+}
+
+bool TaggedUnionMemberCountCheck::isCountingEnumLikeName(
+    StringRef Name) const noexcept {
+  if (llvm::any_of(ParsedCountingEnumPrefixes,
+                   [&Name](const StringRef &Prefix) -> bool {
```

Comment:
```suggestion
                   [Name](StringRef Prefix) -> bool {
```

(Nit: `StringRef`s are glorified pointers, they are designed to be cheap to copy.)